File: Assets/Scripts/Players/Schematics/Schematic.cs
using System;
using System.Collections.Generic;
using System.Linq;

using CatFight.AirConsole.Messages;
using CatFight.Data;
using CatFight.Util;

using UnityEngine;

namespace CatFight.Players.Schematics
{
    [Serializable]
    public sealed class Schematic
    {
        public SchematicData SchematicData { get; }

        private readonly Dictionary<int, SchematicSlot> _slots = new Dictionary<int, SchematicSlot>();

        private readonly List<SchematicSlot> _slotList = new List<SchematicSlot>();

        public IReadOnlyCollection<SchematicSlot> Slots => _slotList;

#if UNITY_EDITOR
        [SerializeField]
        [ReadOnly]
        private SchematicSlot[] _debugSlots;
#endif

        [SerializeField]
        [ReadOnly]
        private bool _isConfirmed;

        public bool IsConfirmed { get { return _isConfirmed; } set { _isConfirmed = value; } }

        [SerializeField]
        [ReadOnly]
        private int _filledSlotCount;

        private readonly Player _player;

        public Schematic(Player player, SchematicData data)
        {
            _player = player;
            SchematicData = data;

            foreach(SchematicSlotData slotData in SchematicData.Slots) {
                SchematicSlot slot = SchematicSlotFactory.Create(slotData);
                if(null == slot) {
                    continue;
                }

                _slots.Add(slotData.Id, slot);
                _slotList.Add(slot);
            }

#if UNITY_EDITOR
            _debugSlots = Slots.ToArray();
#endif
        }

        public void Reset()
        {
            IsConfirmed = false;
            _filledSlotCount = 0;

            foreach(SchematicSlot slot in Slots) {
                slot.Clear();
            }
        }

        public bool SetSlot(int slotId, int itemId)
        {
            if(_filledSlotCount >= SchematicData.MaxFilledSlots) {
                Debug.LogWarning($"Player {_player.DeviceId} has max slots filled!");
                return false;
            }

// TODO: error check (also validate the itemId is legit)
            _slots[slotId].ItemId = itemId;
            ++_filledSlotCount;

            PlayerManager.Instance.BroadcastToTeam(_player.Team.Id, new SetSlotMessage
                {
                    slotId = slotId,
                    itemId = itemId
                },
                _player.DeviceId
            );

            return true;

        }

        public void ClearSlot(int slotId)
        {
// TODO: error check
            int itemId = _slots[slotId].ItemId;
            _slots[slotId].Clear();
            --_filledSlotCount;

            PlayerManager.Instance.BroadcastToTeam(_player.Team.Id, new ClearSlotMessage
                {
                    slotId = slotId,
                    itemId = itemId
                },
                _player.DeviceId
            );
        }
    }
}

File: controller/controller/src/models/schematic.ts
import { GameData } from "./data/gamedata";
import { IItem } from "./item";

export class Schematic {

    private _slots: { [id: number]: number };

    private _filledSlots: number;

    public get filledSlots(): number {
        return this._filledSlots;
    }

    private _slotItems: IItem[][];

    public get slotItems(): IItem[][] {
        return this._slotItems;
    }

    public reset(gameData: GameData): void {

        this._slots = gameData.Fighter ? new Array<number>(gameData.Fighter.Schematic.Slots.length).fill(0) : [];
        this._filledSlots = 0;

        this._slotItems = [];
        if(gameData.Fighter) {

            gameData.Fighter.Schematic.Slots.forEach((slot) => {

                const idx: number = this._slotItems.push([]) - 1;
                gameData.getItems(slot.Type).forEach(item => {
                    this._slotItems[idx].push(item);
                });
            });
        }
    }

    public getSlotItems(slotId: number): IItem[] {

        const slotIdx: number = slotId - 1;
        return this._slotItems[slotIdx];
    }

    public isMaxSlotsSet(gameData: GameData): boolean {
        return this.filledSlots >= gameData.Fighter.Schematic.MaxFilledSlots;
    }

    public isSlotSet(slotId: number): boolean {

        const slotIdx: number = slotId - 1;
        return this._slots[slotIdx] !== 0;
    }

    public isSlotItem(slotId: number, itemId: number): boolean {

        const slotIdx: number = slotId - 1;
        return this._slots[slotIdx] === itemId;
    }

    public setSlot(slotId: number, itemId: number): void {

        if(!this.isSlotSet(slotId)) {
            ++this._filledSlots;
        }

        const slotIdx: number = slotId - 1;
        this._slots[slotIdx] = itemId;
    }

    public clearSlot(slotId: number): void {

        if(this.isSlotSet(slotId)) {
            --this._filledSlots;
        }

        const slotIdx: number = slotId - 1;
        this._slots[slotIdx] = 0;
    }
}
